Sync Log Start Offset between shadow and source partition

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Added a new controller locked task that will sync the start offset from the source to the shadow partition.  This is helpful to ensure consistent client behavior between both partitions when fetching offsets at the start of the log.

Added a new config into the protobuf to support specifiying the interval for the task.

Fixes: CORE-13827

TODO:

- [x] improve unit tests

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
k/d/r: Add delete_offsets serde messages

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/d/rpc: Add delete_records to local service

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/d/rpc: Add RPC for delete_offsets

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
k/d/rpc: Add delete_offsets to client

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Wire in kafka data rpc client

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Wire kafka data rpc client into manager

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Wire kafka RPC accessor into link

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl/model: Add prefix trimming config

Added configuration for the prefix trimming task.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Added prefix_trimmer task

Added a controller locked task that will monitor the log start offsets
of source parititions and, if they happen to change, execute a
delete_records on the shadow cluster to trim shadow partitions.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Wire prefix_trimmer task into service

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
proto: Added PartitionPrefixTrimmingOptions

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
admin/sl: Add support for new prefix trimming options

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
dt: Add tests for prefix trimming

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Rebased off of dev to address merge conflicts

bharathv commented on src/v/kafka/data/rpc/serde.h:
```diff
@@ -155,6 +157,82 @@ struct get_offsets_reply
+    auto serde_fields() { return std::tie(err, low_watermark); }
+
+    kafka::error_code err{kafka::error_code::none};
+    kafka::offset low_watermark{-1};
```
nit: doc

bharathv commented on src/v/kafka/data/rpc/serde.cc:
```diff
@@ -43,6 +43,61 @@ produce_request produce_request::share() {
+fmt::iterator delete_records_request::format_to(fmt::iterator it) const {
+    fmt::format_to(it, "{{cmds: [");
+    bool first_topic = true;
+    for (const auto& [topic, partitions] : cmds) {
```
wondering if there is some magic fmt::join(map, limit)..

wen may need to limit the # of entries printed, these can be huge IIUC.

bharathv replied:
> need to limit the # of entries printed,

fyi.

michael-redpanda replied:
I think this is only logged at trace right?  thoughts on an appropriate upper limit?

IoannisRP replied:
not sure about any join stuff, but there is std::views::take(n) to put a max number of elements in the range.

IoannisRP replied:
a combination of join + take could look something like:
```
fmt::iterator delete_records_request::format_to(fmt::iterator it) const {
    const auto render_partition = std::views::transform(
      [](const auto& p) -> std::string {
          return fmt::format("{{ partition: {}, cmd: {} }}", p.first, p.second);
      });

    const auto render_cmd = std::views::transform(
      [&render_partition](const auto& p) -> std::string {
          constexpr const int max_partitions = 5;
          return fmt::format(
            "{{ topic: {}, partitions: [{}] }}",
            p.first,
            fmt::join(
              p.second | std::views::take(max_partitions) | render_partition,
              ", "));
      });

    constexpr const int max_commands = 5;
    return fmt::format_to(
      it,
      "{{cmds: [{}], timeout: {} }}",
      fmt::join(cmds | std::views::take(max_commands) | render_cmd, ", "),
      timeout);
}
```
Note that performance wise these are not identical.
The current version keeps appending to one big string. This version creates lot's of intermediate strings.
But i am not sure how to use join with our custom formatting for the elements involved (without going through dummy types).

bharathv replied:
>I think this is only logged at trace right?

Sometimes we enable trace for debugging and these get linearized somewhere outside of redpanda and crash with an allocation failure.  think we could also use a truncating logger.

michael-redpanda replied:
actually now that I'm checking the start offset of the mirror partition before issuing the command, maybe this is no longer an issue?  I dont' think we're going to see long sets of commands to trim prefixes

michael-redpanda commented:
Force push:

* Fixed issue with build

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73736-->
<!--abf93cca-fb0c-4ae3-8245-a11e79b607a3-->
<!--324ce5a0-e41c-47aa-85ba-eb502a610e57-->
<!--9c7461a0-671c-4396-8ff0-da031a6b55ee-->
<!--acffb290-59f4-42eb-8e4d-c7744b61a9fa-->
<!--title-->
## Retry command for [Build#73736](https://buildkite.com/redpanda/redpanda/builds/73736)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_auto_prefix_trimming@{"source_cluster_spec":{"cluster_type":"redpanda"}}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_auto_prefix_trimming@{"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--dd344311-ef06-4990-9d34-3f830289c7e6-->
<!--TEST RESULTS-->
<!--905ffc6f-09df-471e-801d-f80ca48615e3-->
<!--TEST RESULTS-->
<!--8bf6f881-a712-448d-b521-ed87b0212fb1-->
<!--TEST RESULTS-->
<!--f8025671-4dc6-4d10-8c46-e8feb9f9766f-->
<!--TEST RESULTS-->
<!--2b81cfe8-05bb-4456-9af6-ece703951ed1-->
<!--TEST RESULTS-->
<!--6a328a77-fc5a-4640-8e6d-89f22e705065-->
<!--TEST RESULTS-->
<!--472c9d5d-b2b3-477d-915b-5a0e0b0fa68c-->
<!--TEST RESULTS-->
<!--e0682cf5-2648-4597-a356-0900c9f54a05-->
<!--TEST RESULTS-->
<!--12d2b024-9d47-48e7-9ad5-329c72164924-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73736">build#73736</a></summary>


| test_class                      | test_method               | test_arguments                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                           |
|:--------------------------------|:--------------------------|:---------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests   | test_auto_prefix_trimming | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73736#0199c060-3a40-4f6e-b6e5-9c02c188711d | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming  |
| ShadowLinkingReplicationTests   | test_auto_prefix_trimming | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73736#0199c066-fbaa-4708-ac0c-6a3d3f890a06 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming  |
| ShadowLinkingReplicationTests   | test_auto_prefix_trimming | {"source_cluster_spec": {"cluster_type": "redpanda"}}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/73736#0199c060-3a42-430d-89fa-6976a9062383 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming  |
| ShadowLinkingReplicationTests   | test_auto_prefix_trimming | {"source_cluster_spec": {"cluster_type": "redpanda"}}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/73736#0199c066-fbac-4ab1-8556-793f6df71de2 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                 | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming  |
| RedpandaNodeOperationsSmokeTest | test_node_ops_smoke_test  | {"cloud_storage_type": 1, "mixed_versions": true}                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/73736#0199c066-fbac-4ab1-8556-793f6df71de2 | FLAKY         | 19/21    | upstream reliability is '98.57142857142858'. current run reliability is '90.47619047619048'. drift is 8.09524 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73754">build#73754</a></summary>


| test_class                 | test_method                    | test_arguments                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                     | test_history                                                                                                                                                            |
|:---------------------------|:-------------------------------|:--------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestKafkaApi       | test_no_auth_enabled           | {"audit_transport_mode": "rpc"} | integration | https://buildkite.com/redpanda/redpanda/builds/73754#0199c15e-ee36-4a75-a30d-3d94349067ec | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled                 |
| ConsumerGroupBalancingTest | test_coordinator_nodes_balance | null                            | integration | https://buildkite.com/redpanda/redpanda/builds/73754#0199c156-ec1b-42bc-894e-fe45d337d9dd | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ConsumerGroupBalancingTest&test_method=test_coordinator_nodes_balance |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73834">build#73834</a></summary>


| test_class                    | test_method            | test_arguments                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                       |
|:------------------------------|:-----------------------|:--------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests | test_replication_basic | {"shuffle_leadership": false, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73834#0199c765-1326-41c8-b24b-330e58537129 | FLAKY         | 20/21    | upstream reliability is '95.74742268041237'. current run reliability is '95.23809523809523'. drift is 0.50933 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73993">build#73993</a></summary>


| test_class                       | test_method                | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                | test_history                                                                                                                                                              |
|:---------------------------------|:---------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest            | test_mount_inexistent      | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73993#0199cf94-8ea7-4f97-bbee-237784d27be9 | FLAKY         | 20/21    | upstream reliability is '99.3576017130621'. current run reliability is '95.23809523809523'. drift is 4.11951 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_mount_inexistent                 |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss | null             | integration | https://buildkite.com/redpanda/redpanda/builds/73993#0199cf95-8bcf-49c5-bc66-ff9c373fd914 | FLAKY         | 18/21    | upstream reliability is '90.3183023872679'. current run reliability is '85.71428571428571'. drift is 4.60402 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74090">build#74090</a></summary>


| test_class              | test_method                                  | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                             |
|:------------------------|:---------------------------------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest   | test_creating_and_listing_migrations         | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/74090#0199deaa-4e77-494b-a715-66b3f9ef734a | FLAKY         | 19/21    | upstream reliability is '98.57142857142858'. current run reliability is '90.47619047619048'. drift is 8.09524 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations |
| MountUnmountIcebergTest | test_simple_remount                          | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/74090#0199deaf-1081-451f-a2de-42ee23bf6bf1 | FLAKY         | 20/21    | upstream reliability is '99.56331877729258'. current run reliability is '95.23809523809523'. drift is 4.32522 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                |
| SegmentMsTest           | test_segment_rolling_with_retention_consumer | null                      | integration | https://buildkite.com/redpanda/redpanda/builds/74090#0199deaa-4e81-4b3a-a064-fb99896ca2cb | FLAKY         | 18/21    | upstream reliability is '91.84993531694697'. current run reliability is '85.71428571428571'. drift is 6.13565 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74202">build#74202</a></summary>


| test_class                    | test_method                     | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                | test_history                                                                                                                                                          |
|:------------------------------|:--------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkBasicTests          | test_rapid_shadow_link_toggling | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/74202#0199e494-42ba-4d20-8df4-a0e79cf87301 | FLAKY         | 19/21    | upstream reliability is '98.4126984126984'. current run reliability is '90.47619047619048'. drift is 7.93651 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkBasicTests&test_method=test_rapid_shadow_link_toggling    |
| ShadowLinkingReplicationTests | test_auto_prefix_trimming       | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/74202#0199e494-42be-499e-b091-09e8f315ec3a | FLAKY         | 19/21    | The test was found to be new, and no failures are allowed                                                                                                             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74246">build#74246</a></summary>


| test_class                   | test_method          | test_arguments                  | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                  |
|:-----------------------------|:---------------------|:--------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestKafkaApi         | test_no_auth_enabled | {"audit_transport_mode": "rpc"} | integration | https://buildkite.com/redpanda/redpanda/builds/74246#0199e7c3-076e-4337-9f57-188ecb9db549 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AuditLogTestKafkaApi&test_method=test_no_auth_enabled       |
| PartitionMovementUpgradeTest | test_basic_upgrade   | null                            | integration | https://buildkite.com/redpanda/redpanda/builds/74246#0199e7c3-0768-486f-ba80-bcee85e2aa55 | FLAKY         | 20/21    | upstream reliability is '99.36974789915966'. current run reliability is '95.23809523809523'. drift is 4.13165 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementUpgradeTest&test_method=test_basic_upgrade |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74278">build#74278</a></summary>


| test_class                              | test_method                  | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                     |
|:----------------------------------------|:-----------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| InternalTopicProtectionLargeClusterTest | test_consumer_offset_topic   | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74278#0199e969-d3a4-44ca-b035-cdcadf7f7ee9 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=InternalTopicProtectionLargeClusterTest&test_method=test_consumer_offset_topic |
| QuickTerminateTest                      | test_terminate               | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74278#0199e969-d39a-4169-b1df-c3d92ac08db6 | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=QuickTerminateTest&test_method=test_terminate                                  |
| TimeQueryTest                           | test_timequery_with_local_gc | null             | integration | https://buildkite.com/redpanda/redpanda/builds/74278#0199e969-d39e-4d47-a66f-ce72a1ff6782 | FLAKY         | 20/21    | upstream reliability is '90.11473962930273'. current run reliability is '95.23809523809523'. drift is -5.12336 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TimeQueryTest&test_method=test_timequery_with_local_gc                         |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/74353">build#74353</a></summary>


| test_class               | test_method                                             | test_arguments                                                        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                                   |
|:-------------------------|:--------------------------------------------------------|:----------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest | test_decommissioning_cancel_ongoing_movements           | {"cloud_topic": true}                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/74353#0199ef8d-36d3-4dd2-a413-a95b6714fc9a | FLAKY         | 14/21    | upstream reliability is '100.0'. current run reliability is '66.66666666666666'. drift is 33.33333 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_cancel_ongoing_movements           |
| NodesDecommissioningTest | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": true}                           | integration | https://buildkite.com/redpanda/redpanda/builds/74353#0199ef8d-36cb-4f27-a560-53c4b9332dbd | FLAKY         | 11/21    | upstream reliability is '89.92805755395683'. current run reliability is '52.38095238095239'. drift is 37.54711 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| NodesDecommissioningTest | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": true}                           | integration | https://buildkite.com/redpanda/redpanda/builds/74353#0199ef91-7640-49cc-922e-8fd8d66d4173 | FLAKY         | 13/21    | upstream reliability is '89.92805755395683'. current run reliability is '61.904761904761905'. drift is 28.0233 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| NodesDecommissioningTest | test_decommissioning_working_node                       | {"cloud_topic": true, "delete_topic": false, "tick_interval": 5000}   | integration | https://buildkite.com/redpanda/redpanda/builds/74353#0199ef91-7647-47cc-b0aa-f74fc6139547 | FLAKY         | 10/21    | upstream reliability is '82.35294117647058'. current run reliability is '47.61904761904761'. drift is 34.73389 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_working_node                       |
| NodesDecommissioningTest | test_decommissioning_working_node                       | {"cloud_topic": true, "delete_topic": true, "tick_interval": 3600000} | integration | https://buildkite.com/redpanda/redpanda/builds/74353#0199ef8d-36c7-4a03-804b-27d80040a9e1 | FLAKY         | 14/21    | upstream reliability is '83.60655737704919'. current run reliability is '66.66666666666666'. drift is 16.93989 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_working_node                       |
| NodesDecommissioningTest | test_decommissioning_working_node                       | {"cloud_topic": true, "delete_topic": true, "tick_interval": 3600000} | integration | https://buildkite.com/redpanda/redpanda/builds/74353#0199ef91-763e-4961-ad5a-3e8b13825371 | FLAKY         | 15/21    | upstream reliability is '83.60655737704919'. current run reliability is '71.42857142857143'. drift is 12.17799 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_working_node                       |
| NodesDecommissioningTest | test_decommissioning_working_node                       | {"cloud_topic": true, "delete_topic": true, "tick_interval": 5000}    | integration | https://buildkite.com/redpanda/redpanda/builds/74353#0199ef8d-36ce-4757-a5ab-0936f4c9a5c0 | FLAKY         | 12/21    | upstream reliability is '90.2439024390244'. current run reliability is '57.14285714285714'. drift is 33.10105 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_working_node                       |
| NodesDecommissioningTest | test_multiple_decommissions                             | {"cloud_topic": true}                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/74353#0199ef91-763d-4bbd-a988-9657d7410765 | FLAKY         | 10/21    | upstream reliability is '64.0'. current run reliability is '47.61904761904761'. drift is 16.38095 and the allowed drift is set to 50. The test should PASS              | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_multiple_decommissions                             |
| NodesDecommissioningTest | test_recommissioning_node_finishes                      | {"cloud_topic": true}                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/74353#0199ef8d-36c6-47dc-9d53-1a05b0de55f3 | FLAKY         | 14/21    | upstream reliability is '100.0'. current run reliability is '66.66666666666666'. drift is 33.33333 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_node_finishes                      |
| NodesDecommissioningTest | test_recommissioning_node_finishes                      | {"cloud_topic": true}                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/74353#0199ef91-763d-4bbd-a988-9657d7410765 | FLAKY         | 12/21    | upstream reliability is '100.0'. current run reliability is '57.14285714285714'. drift is 42.85714 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_recommissioning_node_finishes                      |
</details>
<!--/body-->

michael-redpanda commented:
Force push:

* Added missing `@cluster(num_nodes=8)` that must have been deleted during the rebase

bharathv commented on src/v/kafka/data/rpc/client.cc:
```diff
@@ -429,4 +430,127 @@ client::get_remote_partition_offsets(
+    chunked_hash_map<model::node_id, delete_records_cmd_map> per_node_cmds;
+
+    for (auto& [topic, partitions] : cmds) {
+        co_await ssx::async_for_each_counter(
```
does this benefit from an async_for_each_counter() if the body is not a coroutine.

michael-redpanda replied:
I thought `async_for_each_counter` was built for loops over synchronous code to allow for yields so we don't stall the reactor?

bharathv replied:
ya scratch that, brain freeze.

bharathv commented on src/v/cluster_link/model/types.h:
```diff
@@ -429,6 +429,33 @@ struct topic_metadata_mirroring_config
+    /// Interval for the prefix trimming task
+    std::optional<ss::lowres_clock::duration> task_interval;
+    // Default interval (5 seconds)
+    static constexpr auto task_interval_default = std::chrono::seconds(5);
```
I think this can be lower (30s).

michael-redpanda replied:
higher interval, lower frequency? :)

bharathv replied:
right thats what I meant (especially since prefix trim is such a rare operation)), I 'd even set it to 5 mins or something :))

michael-redpanda replied:
I'm not sure I want to increase the interval... it may have a significant impact on client behavior (if they're doing historical reads or whatever).  I think because it's data related I'd prefer something tighter.

bharathv commented on tests/rptest/tests/cluster_linking_e2e_test.py:
```diff
@@ -699,6 +699,76 @@ def test_replication_with_transactions(self):
+            SecondaryClusterSpec(
+                ServiceType.KAFKA, kafka_version="3.8.0", kafka_quorum="COMBINED_KRAFT"
+            ),
+        ]
```
I'd add a few failures on source / sink to test error paths in trimmer.

bharathv replied:
missed this?

bharathv reviewed (commented):
lgtm.. mostly minor comments.

joe-redpanda commented on src/v/cluster_link/prefix_trimmer.cc:
```diff
@@ -0,0 +1,329 @@
+    co_await cluster.request_metadata_update();
+
+    // Build the list offsets request
+    auto topics = mirror_topics.value() | std::views::keys
```
~~do we want to filter this to only topics with deletion enabled?~~
edit: prefix trim command

joe-redpanda commented on src/v/cluster_link/prefix_trimmer.cc:
```diff
@@ -0,0 +1,329 @@
+  chunked_vector<::model::topic> topics,
+  kafka::client::topic_cache& topic_cache) {
+    chunked_hash_map<::model::node_id, kafka::list_offsets_request> requests;
+    const auto insert_new_node = [&requests](::model::node_id node_id) {
```
nit:
insert_new_node -> add_or_find, and just have it add if not present + return an iterator

joe-redpanda commented on src/v/cluster_link/prefix_trimmer.cc:
```diff
@@ -0,0 +1,329 @@
+            partition_req.current_leader_epoch
+              = topic_cache
+                  .leader_epoch({t, ::model::partition_id{partition_id}})
+                  .value_or(kafka::leader_epoch{-1});
```
nit: kafka::invalid_leader_epoch

michael-redpanda commented:
Force push:

* Addressed PR comments

michael-redpanda commented:
Force push:

* FIxed some failing tests

bharathv commented on src/v/kafka/data/rpc/serde.cc:
```diff
@@ -43,6 +43,61 @@ produce_request produce_request::share() {
+        if (!first_topic) {
+            fmt::format_to(it, ", ");
+        }
+        first_topic = false;
+        fmt::format_to(it, "{{ topic: {}, partitions: [", topic);
+        bool first_partition = true;
+        for (const auto& [partition_id, cmd] : partitions) {
+            if (!first_partition) {
+                fmt::format_to(it, ", ");
+            }
+            first_partition = false;
+            fmt::format_to(
+              it, "{{ partition: {}, cmd: {} }}", partition_id, cmd);
+        }
+        fmt::format_to(it, "] }}");
+    }
+    return fmt::format_to(it, "], timeout: {} }}", timeout);
+}
+
+fmt::iterator delete_records_reply::format_to(fmt::iterator it) const {
+    fmt::format_to(it, "{{ results: [");
+    bool first_topic = true;
+    for (const auto& [topic, partitions] : results) {
+        if (!first_topic) {
+            fmt::format_to(it, ", ");
+        }
+        first_topic = false;
+        fmt::format_to(it, "{{ topic: {}, partitions: [", topic);
+        bool first_partition = true;
+        for (const auto& [partition_id, result] : partitions) {
+            if (!first_partition) {
+                fmt::format_to(it, ", ");
+            }
+            first_partition = false;
+            fmt::format_to(
+              it, "{{ partition: {}, result: {} }}", partition_id, result);
+        }
+        fmt::format_to(it, "] }}");
+    }
+    return fmt::format_to(it, "] }}");
```
this looks like a good suggestion :)

bharathv commented on src/v/cluster_link/prefix_trimmer.cc:
```diff
@@ -0,0 +1,334 @@
+    // offset
+    // 5. If the returned log start offset is greater than the current log start
+    // offset, then the task will issue a prefix trim request to the local
+    // shadow partition
```
this is not what is happening, right? (or am I missing something).

michael-redpanda replied:
no you're right - we'd need to get the partitions starting offset to get that information

michael-redpanda committed:
k/d/rpc: Returning log start offset

Added log start offset to the get_offsets response

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Added `log_start_offset` to the `get_partition_offset` kafka rpc response
* Fetching offsets using kafka rpc to check whether or not we need to trim prefix
* improved testing

IoannisRP commented on src/v/kafka/data/rpc/service.cc:
```diff
@@ -204,6 +205,62 @@ ss::future<result<model::offset, cluster::errc>> local_service::produce(
 
+ss::future<delete_records_result_map> local_service::delete_records(
+  delete_records_cmd_map cmds, model::timeout_clock::duration timeout) {
+    static constexpr int concurrency_limit = 32;
```
where does this `32` come from? I see we use it at `get_offsets` as well.

michael-redpanda replied:
seems to be a standard number used elsewhere

IoannisRP replied:
magic numbers. My favorite :smile:

IoannisRP commented on src/v/kafka/data/rpc/client.cc:
```diff
@@ -429,4 +430,127 @@ client::get_remote_partition_offsets(
+
+    return results;
+}
+delete_records_cmd_map make_copy(const delete_records_cmd_map& cmds) {
+    delete_records_cmd_map result;
+    result.reserve(cmds.size());
+    for (const auto& [topic, partitions] : cmds) {
+        auto& part_copy = result[topic];
+        part_copy.reserve(partitions.size());
+        for (const auto& [partition, cmd] : partitions) {
+            part_copy[partition] = cmd;
+        }
+    }
+    return result;
+}
```
isn't this the default copy constructor?
what am i missing?

michael-redpanda replied:
`delete_records_cmd_map` is a `chunked_hash_map` which does not have a copy constructor

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73954-->
<!--673ba9d6-185a-4e52-a568-2f95a4c4a981-->
<!--title-->
## Retry command for [Build#73954](https://buildkite.com/redpanda/redpanda/builds/73954)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests._perform_auto_prefix_trimming_test
```
<!--/body-->

IoannisRP commented on src/v/kafka/data/rpc/client.cc:
```diff
@@ -429,4 +430,127 @@ client::get_remote_partition_offsets(
+
+    co_await ss::parallel_for_each(per_node_cmds, [this, &results](auto& pair) {
+        auto [node_id, cmds] = std::move(pair);
+        if (node_id == _self) {
+            return _local_service->local()
+              .delete_records(std::move(cmds), timeout)
+              .then([&results](delete_records_result_map rmap) {
+                  join_maps(results, std::move(rmap));
+              });
+        }
+        return remote_delete_records(node_id, make_copy(cmds))
+          .then([node_id, cmds = std::move(cmds)](
+                  result<delete_records_result_map, cluster::errc> res) {
+              if (res.has_error()) {
+                  vlog(
+                    log.warn,
+                    "Failed to delete records on node {}: {}",
+                    node_id,
+                    res.error());
+                  return make_error_results(
+                    cmds, map_topic_error_code(res.error()));
+              }
+              return std::move(res.value());
+          })
+          .then([&results](delete_records_result_map rmap) {
+              join_maps(results, std::move(rmap));
+          });
```
the local `delete_records` seems to be missing the error handling path

michael-redpanda replied:
you mean the `res.has_error()` check?  Local delete records just returns a `delete_records_result_map`.  The remote call returns a result<T, cluster::errc>

IoannisRP replied:
oh yes, my bad. I was looking at the other `delete_records`. All good.

michael-redpanda commented:
Force push:

* renamed method

IoannisRP commented on src/v/cluster_link/prefix_trimmer.cc:
```diff
@@ -0,0 +1,395 @@
+        vlog(logger().warn, "{}", msg);
+        std::ignore = change_state(model::task_state::link_unavailable, msg);
+        co_return;
+    }
```
it feels like you are using the `runtime_error` exceptions for flow control. Is the catch meant to capture only these 2?

michael-redpanda replied:
yeah so I didn't repeat the create msg, log, change state code.  I can if you think that's better

IoannisRP replied:
my main concern was that the catch might intercept a different exception, which may or may not be a problem.

I think the new version is better :+1: 
For the duplication, you could add a lambda where you just pass it the error msg, but i think it's ok as is.

IoannisRP commented on src/v/cluster_link/prefix_trimmer.cc:
```diff
@@ -0,0 +1,395 @@
+chunked_hash_map<::model::node_id, kafka::list_offsets_request>
+build_list_offsets_request(
+  const chunked_vector<::model::topic>& topics,
+  kafka::client::topic_cache& topic_cache) {
```
```suggestion
  const kafka::client::topic_cache& topic_cache) {
```

IoannisRP commented on src/v/cluster_link/prefix_trimmer.cc:
```diff
@@ -0,0 +1,395 @@
+
+chunked_vector<kafka::data::rpc::topic_partitions> build_get_offsets_request(
+  const chunked_vector<::model::topic>& topics,
+  kafka::client::topic_cache& topic_cache) {
```
```suggestion
  const kafka::client::topic_cache& topic_cache) {
```

IoannisRP commented on src/v/cluster_link/prefix_trimmer.cc:
```diff
@@ -0,0 +1,395 @@
+                node_topics.emplace(
+                  *leader, kafka::list_offset_topic{.name = t});
+            }
+            node_topics[*leader].partitions.push_back(partition_req);
```
nit: there is a double search here. Could be reduced to a single at a slight cost to readability.

michael-redpanda commented:
Force push:

* Addressed PR comments

IoannisRP reviewed (dismissed):
lgtm

michael-redpanda commented:
Force push:

* Rebased off of dev

michael-redpanda commented:
Force push:

* Rebased off of dev

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/74202-->
<!--4f56ffa0-0737-41c6-8bae-39eb85da1c6b-->
<!--title-->
## Retry command for [Build#74202](https://buildkite.com/redpanda/redpanda/builds/74202)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_auto_prefix_trimming@{"source_cluster_spec":{"cluster_type":"redpanda"},"with_failures":true}
```
<!--/body-->

michael-redpanda commented:
Force push:

* Rebased off of dev

michael-redpanda commented:
Force push:

* Kafka data RPC client now is always created - shadow linking depends on it to always be there

rockwotj commented on proto/redpanda/core/admin/v2/shadow_link.proto:
```diff
@@ -281,6 +283,14 @@ message SecuritySettingsSyncOptions {
     // ACL filters
     repeated ACLFilter acl_filters = 5;
 }
+// Options for the prefix trimmer task
```
Please explain what a prefix trimmer is. Anybody reading this will not know

rockwotj replied:
Checkout the doc team's guide on comments.

https://redpandadata.atlassian.net/wiki/spaces/CORE/pages/1360068616/Admin+API+v2#Other-writing-tips

rockwotj reviewed (commented):
Protos LGTM except for the comments should explain the feature and what it does and why someone cares

michael-redpanda commented:
Force push:

* Addressed comments on proto

michael-redpanda commented:
Will remove ListOffset use and use the start offset from fetcher once #27976 merges

michael-redpanda commented:
Force push:

* Rebased off of dev

michael-redpanda committed:
cl: Add method to task to handle partition leadership changes

Adds a new virtual method to the task class that can be overridden by
implementations to handle when partition leadership changes.  This will
be called by the link running the task.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cl: Wire in task notification of leadership changes

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Using start offset from fetcher rather than ListOffsets

michael-redpanda commented:
Rebased off of dev

bharathv commented on src/v/cluster_link/prefix_trimmer.cc:
```diff
@@ -0,0 +1,246 @@
+}
+} // namespace
+
+prefix_trimmer::prefix_trimmer(link* link, const model::metadata& config)
+  : task(
+      link,
+      config.configuration.partition_prefix_trimming_cfg.get_task_interval(),
+      prefix_trimmer::task_name)
+  , _config(config.configuration.partition_prefix_trimming_cfg) {}
+
+void prefix_trimmer::update_config(const model::metadata& config) {
+    _config = config.configuration.partition_prefix_trimming_cfg;
+    set_run_interval(
```
I'm a bit confused.. do we need any of these "task" changes?  "start offsets" are now available in the replicators (thanks to you status API changes), so they are also available in the local_partition_sink (which can just trigger an prefix trim action if there is an offset mismatch?)

paulohtb6 commented on proto/redpanda/core/admin/v2/shadow_link.proto:
```diff
@@ -338,6 +340,18 @@ message SecuritySettingsSyncOptions {
+// offset has changed, then this task will issue a prefix-trim operation against
+// the mirror partition to keep it in sync.
+message PartitionPrefixTrimmingOptions {
+    // Whether or not the prefix trimmer is enabled
```
```suggestion
    // Whether the prefix trimmer is enabled
```

paulohtb6 commented on proto/redpanda/core/admin/v2/shadow_link.proto:
```diff
@@ -338,6 +340,18 @@ message SecuritySettingsSyncOptions {
     // ACL filters
     repeated ACLFilter acl_filters = 5;
 }
+// The following options are for the partition prefix trimmer task.  This task
+// is used to keep the log starting offset of shadow partitions in sync with the
+// partitions on the source cluster.  If the task detects that the starting
+// offset has changed, then this task will issue a prefix-trim operation against
+// the mirror partition to keep it in sync.
```
```suggestion
// The following options configure the partition prefix trimmer task. This task
// keeps the log starting offset of shadow partitions synchronized with 
// partitions on the source cluster. If the task detects that the starting
// offset changed, it issues a prefix-trim operation against the shadow partition
// to maintain synchronization.
```

paulohtb6 commented on proto/redpanda/core/admin/v2/shadow_link.proto:
```diff
@@ -338,6 +340,18 @@ message SecuritySettingsSyncOptions {
+    // Whether or not the prefix trimmer is enabled
+    bool enabled = 1;
+    // How often to run the prefix trimmer
+    // If 0 provided, defaults to 5 seconds
```
```suggestion
    // Defaults to 5 seconds if set to 0
```

michael-redpanda commented:
@tyson-redpanda - i'm taking this off the rc1 candidate.  This PR may close and a new one (that's much simpler and doesn' thave protobuf changes) will come up

michael-redpanda commented:
Closed in favor of https://github.com/redpanda-data/redpanda/pull/28142
